yktrace/src/hwt/mapper.rs

```diff
@@ -108,6 +108,61 @@ impl HWTMapper {
         }
     }
 
+    /// Collapse consecutive control point blocks (including any unmappable internals) into one
+    /// control point entry. Unmappable holes exist because the generated control point calls out to
+    /// Rust code for which we have no IR.
+    ///
+    /// For example, if the trace is:
+    ///
+    /// ```ignore
+    /// [bb1, bb2, bb3, ctrlp, ctrlp, ?, ctrlp, bb4]
+    /// ```
+    ///
+    /// (where `bbN` is a mappable block, `ctrlp` is a block in the control point, and `?` is
+    /// unmappable blocks), then this function would canonicalise the trace to:
+    ///
+    /// ```ignore
+    /// [bb1, bb2, bb3, ctrlp, bb4]
+    /// ```
+    fn canonicalise_control_point(irblocks: &mut Vec<IRBlock>) {
+        let ctrlp_sym_c = CString::new(CONTROL_POINT_SYM).unwrap();
+        assert!(irblocks[0].func_name == ctrlp_sym_c);
+        let mut in_ctrlp = true;
+        let mut ctrlp_start_idx = Some(0);
+        let mut delete_ranges = Vec::new();
```

Comment:
Reading the code below it feels like it might be simpler to copy the "keep" things rather than later deleting the "don't keep" things? I might be wrong!

Reply:
I was trying to do it copy-free, but if you think it's not worthwhile, I could probably simplify the code a bit.

Reply:
Deleting a vec's elements does do copying. However, I don't feel strongly: you should write this whichever you way you think is clearer.

Reply:
Ah, yes, shuffling the elements down...

Let's try a rewrite to see if it looks cleaner.

Reply:
8470f3b7 (and 79b6ef67) is the most concise I managed to get it. What do you think?

[It seems tempting to chain the first two `if`s but you end up in a pickle it turns out]

Reply:
Works for me!